Bluetooth: Mesh: Persistent storage of Virtual Addresses

LingaoM opened the pull request:
The 16-bit format group addresses will be stored,
but we don't store (or restore) the virtual label UUIDs,
i.e. after a power cycle the 16-bit group addresses
would be meaningless.

Fixed zephyrproject-rtos#19342

Signed-off-by: Lingao Meng <mengabc1086@gmail.com>

LingaoM committed:
Bluetooth: Mesh: Persistent storage of Virtual Addresses

The 16-bit format group addresses will be stored,
but we don't store (or restore) the virtual label UUIDs,
i.e. after a power cycle the 16-bit group addresses
would be meaningless.

Fixed zephyrproject-rtos#19342

Signed-off-by: Lingao Meng <mengabc1086@gmail.com>

trond-snekvik commented on subsys/bluetooth/mesh/cfg_srv.c:
```diff
@@ -1130,6 +1162,11 @@ static u8_t va_add(u8_t *label_uuid, u16_t *addr)
+
+	if (IS_ENABLED(CONFIG_BT_SETTINGS)) {
+		bt_mesh_store_va();
+	}
```
Consider putting this pattern in a function

trond-snekvik reviewed (changes requested):
Looks like a good solution to me overall :+1:

jhedberg commented on subsys/bluetooth/mesh/settings.c:
```diff
@@ -1354,6 +1401,54 @@ static void store_pending_mod(struct bt_mesh_model *mod,
+		return;
+	}
+
+	snprintk(path, sizeof(path), "bt/mesh/Va/%x", l->addr);
```
> that's an issue other places in the stack as well

Where? At least for incoming messages the full virtual label UUID is used as part of the decryption process, which would fail if the 16-bit version had false-positive match. I.e. the message wouldn't be passed up to the model layer in that case.

trond-snekvik replied:
The incoming messages are only decrypted for the first virtual label UUID that matches the 16 bit address, but if two virtual label UUIDs share the same 16 bit address, we'll never try decrypting with the second. 

Chances of this are pretty microscopic, but it's very hard to control and detect the issue for users, as the 16 bit address part of the UUID is an internal detail in the transport layer. The current problem can be fixed later without too much hassle, but if we start doing this in the persistent storage, it becomes a more permanent problem.

jhedberg replied:
@trond-snekvik agreed, thanks for elaborating

LingaoM committed:
Bluetooth: Mesh: Persistent storage of Virtual Addresses

The 16-bit format group addresses will be stored,
but we don't store (or restore) the virtual label UUIDs,
i.e. after a power cycle the 16-bit group addresses
would be meaningless.

Fixed zephyrproject-rtos#19342

Signed-off-by: Lingao Meng <mengabc1086@gmail.com>

LingaoM committed:
Update cfg_srv.c

LingaoM committed:
Bluetooth: Mesh: Move out Public Header

This is not a public structure.
It should be placed inside rather than exposed.

Signed-off-by: Lingao Meng <mengabc1086@gmail.com>

LingaoM committed:
Bluetooth: Mesh: Fixed Address Conflict Problem

We will modify the index according to address storage to
solve the potential address conflict problem.

Signed-off-by: Lingao Meng mengabc1086@gmail.com

LingaoM committed:
Bluetooth: Mesh: Fixed Address Conflict Problem

We will modify the index according to address storage to
solve the potential address conflict problem.

Signed-off-by: Lingao Meng mengabc1086@gmail.com

LingaoM committed:
Bluetooth: Mesh: Fix potential null pointer problems

Fix potential null pointer problems when get NULL return.

Signed-off-by: Lingao Meng mengabc1086@gmail.com

LingaoM committed:
Bluetooth: Mesh: Add index params.

Add index parameters. Easy to delete lables.

Signed-off-by: Lingao Meng mengabc1086@gmail.com

LingaoM committed:
Bluetooth: Mesh: Fixed Address Conflict Problem

Modify the prototype of some functions to solve the
potential address conflict problem.

Signed-off-by: Lingao Meng mengabc1086@gmail.com

LingaoM committed:
Bluetooth: Mesh: Fixed Address Conflict Problem

Modify the prototype of some functions to solve the
potential address conflict problem.

Signed-off-by: Lingao Meng mengabc1086@gmail.com

LingaoM committed:
Bluetooth: Mesh: Add return parameters

Instead of keeping the default and warning,
I should show the instructions return value.

Signed-off-by: Lingao Meng mengabc1086@gmail.com

trond-snekvik reviewed (commented):
LGTM, apart from a couple of trivial issues and the checkpatch problems.

To silence the git log warnings, you need to edit your commit messages to use the correct sign-off format, and add a body to 6086d0f. I would do it like this:

`git rebase -i 6086d0f -x 'git commit --amend -s'`

This'll first bring up a list of all the commits from 6086d0f to HEAD, with an action in front of them, and the execute line between each. Just save and exit this, and your commit messages will be opened in your editor one after one, with the correct sign-off footer. Make sure you delete any duplicate sign-off messages (keep the ones with brackets around the email address!). Afterwards, check that your log looks good, then force push to your branch. 

The bot complains because some of your sign-off lines are missing the `<>` brackets around your email address. For any new commits, git will append a correctly formatted sign-off line automatically if you always use the `-s` flag when you commit.

LingoMeng commented:
I'm close this PR, because a more concise alternative can be used. see PR #19799
